### ISSUE TITLE
Settings could be managed network wide if the plugin is network activated.

### ISSUE THREAD

**Issue by thomaslhotta** [bcole808/wp-social-metrics-tracker#69]
If the plugin is activated network wide it would only make sense to allow the options to be administrated from the network admin. 

I understand there may be some issues, such as that different post types may be available on different blogs. In my case I do not want to give the admins of individual blogs the ability to change the tracking settings and all blogs share the same post types, so i do not have this issue. Maybe the options could only be moved to the network admin if a specific constant is set or something like that. 

If anyone is interested in such a feature I could create a pull request.

**Comment by bcole808** [comment]
Hello, I am actually also in need of this feature myself since I use this plugin primarily on Multisite. Right now I just manually change the settings for each blog (which is less convenient).  I haven't ever created a Multisite plugin settings page though so I don't know what the best way is to implement this. If you would be up for helping with a pull request that would be super awesome! 

Can we discuss here what the best way is to implement this? What are your thoughts? 

# Desired behavior

My thoughts on how this feature should work: If the plugin is activated on Multisite, the network admin should have the option for the plugin to operate in one of these ways:
1. Each site admin can set individual options (default)
2. All options can be set once for the entire network, no individual site can be overridden. 
3. (Maybe, if convenient to develop) All options can be set for the network, and allow individual blogs to be overridden if configured by the admin (maybe useful for custom post types)

# How settings work now:

### Get/Set options for the plugin

Settings are loaded in [social-metrics-tracker.php](https://github.com/chapmanu/wp-social-metrics-tracker/blob/development/src/social-metrics-tracker.php#L85-L88) and are [read/written with set and get methods](https://github.com/chapmanu/wp-social-metrics-tracker/blob/development/src/social-metrics-tracker.php#L312-L353) a lot like regular WP options. It is done this way to save all the key/value pairs in one single wp_options entry in the database. This should make it easy to toggle between `update_option()` and `update_site_option()` if that is the best way to do multisite options. 

### Settings page UI

The [settings page](https://github.com/chapmanu/wp-social-metrics-tracker/blob/development/src/SocialMetricsSettings.class.php) is a little bit messy in terms of how it is coded. The settings page is divided into "tabs" and each one works a little bit differently. Most often, the settings are saved using the plugins `update_smt_option()` as referenced above, but once in a while it writes an option directly using WordPresses `update_option`. Not sure if this can be easily used on the Network options area or not, it would be great if the same code could be re-used. 

Thoughts?

**Comment by thomaslhotta** [comment]
Hi. I completely agree with the desired behaviour you outlined. I have created a few multisite plugins for my personal use, but never any public ones. So I don't have much experience with best practices in implementing network admin settings pages and so on. However there are a few popular plugins out there that could be used as an example.

First I think we would need a network admin settings page if the plugin is activated network wide.  There the admin can select the desired behaviour. That should be pretty straight forward to implement. 

For the actual settings page I think there are three ways to approach this:

### Allow one blog to be set as the "master" in the network admin interface.

All other blogs would pull the settings from this blog. In most cases this would be the main blog. An example of a plugin operating in this way would be myCRED. 

**Pros**:
- No changes to the settings page have to be made, only to the way the options are retrieved
- Flexible, any blog could be set as master. 

**Cons**
- Maybe harder to understand for users?

### Move the settings page to the network admin

If behaviour 2 or 3 are selected a settings page is created in the network admin. The WordPress Performance Pack plugin operates this way.

**Pros**:
- Kind of where the settings belong
- No cluttering up the settings menu of the normal admin
- Maybe easier for users to understand

**Cons**:
- Changes to the settings page and CSS have to be made to make it work. I gave this a quick try, but couldn't get everything working without more extensive changes.
- The network admin operates in the context of the main blog. I don't know if this is intentional or just a side effect from the way WordPress determines the current blog. Anyway I'm not sure if functions like `get_post_types` can be trusted in the network admin, even if they do seem to work.

### Allow hard coding the settings in wp-config.php

Anybody running a WordPress multisite will have the skills to modify the config file, so this might be worth considering. This would remove the need for any settings page if the options are hard coded.

**Pros**:
- Very easy to implement
- May better suit the needs of people who manages things like API keys in the server environment config instead of the database

**Cons**:
- Not very user friendly

### About the 3rd behaviour

This could be solved quite simply. Just retrieve the default network wide settings and then merging the array with the settings of the individual blogs. The only place where I see an issue are the API keys for Facebook and Google. If they are set network wide you may not want to reveal them to admins of the individual blogs.

I personally favour option 2 or 3 because they are simpler. But there may be other arguments against them...

**Comment by bcole808** [comment]
My thoughts on these three options below. 

#### 1. Allow one blog to be set as the "master" in the network admin interface.

> Con: Maybe harder to understand for users?

Agreed. So probably not this one. 

#### 2. Move the settings page to the network admin

**This is probably the way most plugin users would expect such a feature to operate.** It is similar to how [W3-Total-Cache](https://wordpress.org/plugins/w3-total-cache/) implements this feature. They offer a check-box in the network admin area that says something like "Allow individual blogs to override these settings if desired" or "Force all blogs to use these settings"

But I also agree that it could be quite complex to implement and would require a lot of code cleanup / changes to the settings page. 

That said, the code for the settings page is messy as-is and does need some cleanup....

#### 3. Allow hard coding the settings in wp-config.php

Agree that this is the fastest, and easiest to implement. Also agree that site admins running multisite will probably be savvy enough to use this feature. 

More thoughts on option three: 
- It would be great if on the actual settings page the individual items it simply became grayed out and the value from the config file was displayed (along with a tooltip stating "this setting is already defined site-wide")
- If a network admin hard-codes an option key, then [that key should probably never change in this plugin](https://github.com/chapmanu/wp-social-metrics-tracker/blob/development/src/settings/smt-general.php) otherwise the custom defined values won't match up. (So maybe this means putting some developer notes in the comments of the plugin warning that option keys should never be changed, unless really important and also documented in the changelog)

# Thoughts

Seems like **Option 2** is the best for the user, but harder to implement. 
Seems like **Option 3** will work too, but is easier to implement. 

Which would you prefer to work on for a pull request? 

It might be good to make a branch for this and send the pull request to the branch. That way we can do plenty of testing and make sure to get it working well before merging back for publishing. I [made a branch here](https://github.com/chapmanu/wp-social-metrics-tracker/tree/issue-69-multisite-settings). (Let's see if this works this way, I haven't tried it before so let me know if there is a better way to work)

**Comment by thomaslhotta** [comment]
So I have been digging into this a bit.

I think we only need two options for the network behaviour: "Allow override" and "don't allow override" of the network settings. The first option ("each site admin can set individual options") is kind of implicitly set if the plugin is not network activated, as this gives the choice to to blog admin to activated the plugin or not.

I also noticed another issue. It seems like the WordPress Settings Framework does not support `get_site_option`. This might need some ugly workaround.

**Comment by bcole808** [comment]
Agreed, the two behaviors can be A) Network settings override all site settings, or B) Individual blog settings override network settings. 

I would be okay with eliminating WordPress Settings Framework, or just taking the good parts and getting rid of the rest.  I started off with that framework, but halfway through making this plugin I started trying to do things with the settings page that the framework wouldn't allow and so I switched away from it. So that's why half of the settings are done with the framework and the newer items are outside of it. 

I've been using [these functions](https://github.com/chapmanu/wp-social-metrics-tracker/blob/development/src/social-metrics-tracker.php#L313-L347) mostly to manage plugin options.
